cmd/frontend/graphqlbackend/search.go

```diff
@@ -510,7 +515,7 @@ func (r *searchResolver) resolveRepositories(ctx context.Context, effectiveRepoF
 
 	settings, err := decodedViewerFinalSettings(ctx)
 	if err != nil {
-		return nil, nil, nil, false, err
+		return resolved, err
```

Comment:
Wherever we return an error, it becomes less clear what `resolved` actually contains. One has to scroll up and check whether `resolved` has been written to. Would it be easier to read if we returned an empty struct (or `nil` if we changed the returned type to a pointer)? What do you think? 
```suggestion
		return resolvedRepositories{}, err
```

Reply:
Yeah that makes sense to me.

Reply:
yeah lets change it to a pointer.

Reply:
Hmm I'm fine with @stefanhengl suggestion just making this a non-nil struct value initialized to nil members. I'm not confident about changing the value to a pointer. If the suggestion is that all return values are initialized as `&resolvedRepositories{}` and we do not return `nil` anywhere, this is OK. But if you mean that returns here should be substituted for nil, I'm not so confident. Like, stylistically, sure, that's nice, but semantically, I can then no longer think through these changes mechanically--I have to then also think about the callers and checking for nil on this value, and in what contexts that's appropriate or not, and possibly miss something. Not saying it can't be done, it just makes me anxious.

Reply:
Normally in go if you return a non-nil error, the first value is nil. If you look at the call sites for this function you should see the same behaviour of ignoring the return value of non-nil returns?

Reply:
> Normally in go if you return a non-nil error, the first value is nil.

Yeah, but this is a sort of "there's the rub" issue... remember [this one](https://github.com/sourcegraph/sourcegraph/pull/12437#pullrequestreview-454406363):

> Maybe we should just make searchFilesInRepos always return an initialized one? Seems a shame to have to check for nil everywhere.

So OK, I can deal with the fact that checking for nil everywhere is the convention. But in the context of a refactor like this, it's not mechanical/straightforward. Because fundamentally, I'm introducing a product type here to hold values that are initialized to zero values. If we allow this type to be either nil or non-nil, we've introduced extra state to consider in each call site context. Some of these rely on, say `overLimit` but don't care about the other values. Checking for each call site where `nil` would be OK is wearisome and error-prone--I might feel different if I knew these paths were well-tested, but I don't since that one time I introduced a panic with nil-derefs in commits search.

I would even speculate that the reason this return grew to seven values is because it's easier to ensure nothing breaks by tacking on more return values than mint a dedicated type (particularly if the convention creates an expectation/obligation to turn it into a `nil` pointer). By minting a dedicated type without turning it into a pointer `nil` returns, I'm more in line with the thinking "don't let the perfect be the enemy of the good". So I think this will improve things despite not being perfectly inline with conventions, while leaving open the possibility to bring it in line with the convention. Sorry for the essay, but it's worth typing out.

Reply:
I filed https://github.com/sourcegraph/sourcegraph/issues/13310 for this, tracking follow up refactoring opportunities.